src/main/groovy/org/xbib/gradle/plugin/JFlexPlugin.groovy

```diff
@@ -52,36 +53,66 @@ class JFlexPlugin implements Plugin<Project> {
         project.extensions.create ('jflex', JFlexExtension)
     }
 
-    private static void addJFlexTasks(Project project) {
-        project.sourceSets.all { SourceSet sourceSet ->
-            addJFlexTaskForSourceSet(project, sourceSet)
+    private static void addJFlexTasks(Project project, JFlexExtension jFlexExtension) {
+        project.extensions.getByType(SourceSetContainer).all { SourceSet sourceSet ->
+            addJFlexTaskForSourceSet(project, sourceSet, jFlexExtension)
         }
     }
 
-    private static void addJFlexTaskForSourceSet(Project project, SourceSet sourceSet) {
+    private static void addJFlexTaskForSourceSet(Project project, SourceSet sourceSet, JFlexExtension jFlexExtension) {
         String taskName = sourceSet.getTaskName('generate', 'jflex')
         SourceDirectorySet sourceDirectorySet = sourceSet.extensions.getByName('jflex') as SourceDirectorySet
-        File targetFile = project.file("${project.buildDir}/generated/sources/${sourceSet.name}")
-        if (sourceDirectorySet.asList()) {
-            TaskProvider<JFlexTask> taskProvider = project.tasks.register(taskName, JFlexTask) {
-                group = 'jflex'
-                description = 'Generates code from JFlex files in ' + sourceSet.name
-                source = sourceDirectorySet.asList()
-                target = targetFile
-                theSourceSet = sourceSet
-            }
-            logger.info "created ${taskName} for sources ${sourceDirectorySet.asList()} and target ${targetFile}"
-            project.tasks.named(sourceSet.compileJavaTaskName).configure({
-                dependsOn taskProvider
-            })
-            if (sourceSet.java && sourceSet.java.srcDirs) {
-                sourceSet.java.srcDirs += targetFile
+        Provider<Directory> defaultTargetFile = project.layout.buildDirectory.dir("generated/sources/${sourceSet.name}")
+
+        Provider<File> targetFile = jFlexExtension.writeIntoJavaSrc.map {
+            if (it) {
+                if (sourceSet.java && sourceSet.java.srcDirs) {
+                    logger.info "java sources: ${sourceSet.java.srcDirs}"
+                    File target = sourceSet.java.srcDirs.first()
+                    logger.info "switching to first java source directory ${target}"
+                    return target
+                } else {
+                    logger.warn "writing into java source not possible, is empty"
+                    return defaultTargetFile.get().asFile
+                }
+            } else {
+                return defaultTargetFile.get().asFile
             }
         }
+        TaskProvider<JFlexTask> taskProvider = project.tasks.register(taskName, JFlexTask) {
+            group = 'jflex'
+            description = 'Generates code from JFlex files in ' + sourceSet.name
+            source = sourceDirectorySet.asList()
+            target.fileProvider(targetFile)
+
+            encoding.set(jFlexExtension.encoding)
```

Comment:
this simply wires up properties from the extension to their corresponding task inputs. Defaults are defined as before on the extension and then passed to the task and resolved at task execution time